tutorials/torchao_vllm_integration.md

```diff
@@ -0,0 +1,465 @@
+
+```python
+@MyQuantizedTensor.implements(torch.nn.functional.linear)
+def _(func, types, args, kwargs):
+    input_tensor, weight_tensor, bias = args[0], args[1], args[2] if len(args) > 2 else None
+
+    # This is where you define what hardware your method supports
+    if hasattr(weight_tensor, 'use_cutlass_kernel'):
+        return my_cutlass_linear(input_tensor, weight_tensor, bias)
+    elif hasattr(weight_tensor, 'use_triton_kernel'):
+        return my_triton_linear(input_tensor, weight_tensor, bias)
+    else:
+        # Fallback - dequantize and use standard linear
+        return torch.nn.functional.linear(
+            input_tensor, weight_tensor.dequantize(), bias
+        )
+```
+
+#### Compilation Benefits
+The overhead of tensor subclasses disappears with `torch.compile()`, this is on by default in VLLM:
+
+```python
+# Without compilation: subclass dispatch overhead
+model = quantize_(model, MyNewQuantConfig())
+
+# With compilation: optimized kernels, no dispatch overhead
+model = torch.compile(quantize_(model, MyNewQuantConfig()))
+```
+
+### Trade Off of Tensor Subclasses
+1. **Compilation**: is essential for removing subclass overhead. Without it unless your model is extremely gpu bound the overhead of dispatch on the CPU can severely impact performance.
+2. The checkpoint defines the behavior of the model. You might be saying "don't all checkpoints do this". This is true, however people typically solely think of a torch.Tensor as its data. When in actuality its a true class where it brings the Dispatcher and all the kernels ATen has registered to it. When you define your tensor subclass, you are building a separate little world. One w/ a different representation of data, but also one where you need to explicitly define what ops you support and have implementations for all the hardware you want to support. This can feel a little like spooky action at a distance at first. But it can be very powerful. Case in point is being able to support TP with only 3 definitions.
+
+## Serialization and Model Sharing
+
+### SafeTensors Limitations
+
+**Important**: TorchAO quantized models currently cannot be serialized with safetensors due to tensor subclass limitations. SafeTensors only supports plain tensors, but TorchAO uses tensor subclasses with additional metadata that cannot be directly serialized.
+
+### Current Workaround
+
+When saving quantized models to HuggingFace Hub, you **must** use `safe_serialization=False`:
+
+```python
+from transformers import TorchAoConfig, AutoModelForCausalLM
+from torchao.quantization import Int4WeightOnlyConfig
+
+# Quantize model
+quantization_config = TorchAoConfig(Int4WeightOnlyConfig(group_size=128, use_hqq=True))
+model = AutoModelForCausalLM.from_pretrained(
+    "meta-llama/Llama-2-7b-hf",
+    quantization_config=quantization_config
+)
+
+# REQUIRED: Must use safe_serialization=False for tensor subclasses
+model.push_to_hub("your-username/llama-2-7b-int4", safe_serialization=False)
```

Comment:
I feel this code block is duplicated from above? Should we just mention `safe_seiralization` must be False up there?